### DISCUSSION TITLE
How can I display [link-id]: https://url.tld "title" links automatically where they're defined?

### DISCUSSION THREAD

**Discussion by ReedyBear** [thephpleague/commonmark#734] [Q&A] [answered]
I want to have something like:
```markdown
my county's population was about [110,000 in 2010][2010-population] and blah blah

then at the bottom

[2010-population]: http://the_url "Census population counts 2010 - 2019"
```

And of course render the link in the body of the article. But then ALSO render a link at the bottom, to have like a ... sources section. But the `[2010-population]: http:// ...` part does not render into html.

I'm doing this to render:
```php
$cmark = new \League\CommonMark\CommonMarkConverter();
$convertedToHtml = $cmark->convertToHtml($md);
```

Are there configuration options, or an extension I can use that would auto-render the id-style-links into the html?

**Comment by ReedyBear** [comment]
Of course, it would be pretty easy to re-use the [link-id] to make a sources section myself ... but I'm lazy & who wants to do that? lol

**Comment by colinodell** [comment, marked as answer]
There are currently two different ways that (kind of) support what you're looking for:

 - Link reference definitions: https://spec.commonmark.org/0.30/#link-reference-definitions
 - Our `FootnotesExtension`: https://commonmark.thephpleague.com/2.0/extensions/footnotes/

But to get the exact functionality you described, you would need to write a small extension.  The heart of that extension would be an event listener for [the `DocumentParsedEvent`](https://commonmark.thephpleague.com/2.0/customization/event-dispatcher/#list-of-available-events) that takes the parsed [`Document`](https://commonmark.thephpleague.com/2.0/customization/abstract-syntax-tree/#document), iterates through all the references inside of [its `ReferenceMap`](https://github.com/thephpleague/commonmark/blob/df50865e23dd6634b1ec441ae59194bdb8171efc/src/Node/Block/Document.php#L37), and [appends](https://commonmark.thephpleague.com/2.0/customization/abstract-syntax-tree/#modification) link nodes (perhaps in a list?) to the bottom of the `Document` node so that they get rendered.  I think that should give you what you're looking for :)

**Reply by ReedyBear** [reply]
Thank you!
